File: Python/112.Path Sum .py
#判断是否存在一条从根节点到叶节点的路径，这条路径的数之和为目标数
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def hasPathSum(self, root, sum):
        self.sum = sum
        return self.inorder(root,0)

    def inorder(self,p,cursum):
        if p!=None:
            cursum += p.val
            if p.left==None and p.right==None:
                return cursum==self.sum
            else:
                if self.inorder(p.left,cursum)==True:
                    return True
                if self.inorder(p.right,cursum)==True:
                    return True
                return False


t1 = TreeNode(5)
t2 = TreeNode(4)
t3 = TreeNode(8)
t4 = TreeNode(11)
t5 = TreeNode(13)
t6 = TreeNode(4)
t7 = TreeNode(7)
t8 = TreeNode(2)
t9 = TreeNode(1)
t1.left = t2
t1.right = t3
t2.left = t4
t3.right = t6
t3.left = t5
t4.left = t7
t4.right = t8
t6.right = t9

s = Solution()
print(s.hasPathSum(t1,19))

File: C++/Tree/112.Path Sum.cpp
/*
判断是否存在一条从根到叶的路径，它的结点和等于给出的数。
*/
/*
思路：向下递归，把当前结点的值减掉传入下一层，到达叶结点时判断是否等于传入的值即可。
*/
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
	bool hasPathSum(TreeNode* root, int sum) {
        if(!root)	return false;
        if(!root->left && !root->right)	return root->val==sum; 
        return hasPathSum(root->left,sum-root->val) || hasPathSum(root->right,sum-root->val);
    }
};
int main()
{
	Solution s;
	TreeNode root(1),t1(2),t2(3);
	root.right = &t1;
	t1.left = &t2;
	cout << s.hasPathSum(&root,6)<<endl;
}

